### PR TITLE
Translate Unstructured objects to concrete instances

### LINKED ISSUE #613: Strategic Patch fails with fake dynamic client
When using the fake client, strategic merge patch operations will fail with the error `unable to find api field in struct Unstructured for the json field "metadata"` if the object being patched was created using the dynamic fake client.

I think that the reason is that fake object tracker stores objects created with the dynamic simple client as `Unstructured` object instances, but when a strategic patch is being applied the strategic patch code assumes that the original object is a concrete instance of whatever is being patched.

Here's a small program that simulates the issue (compiled with the `kubernetes-1.14.0` release of client-go): https://gist.github.com/liorokman/2dd4ac43e1e3fcee7c7d4588178d27d6 

When running against a real Kubernetes cluster the program runs correctly.

When running with the fake clientset, the program crashes:
```
panic: unable to find api field in struct Unstructured for the json field "metadata"

goroutine 1 [running]:
main.main()
	/home/lior/go/src/github.com/liorokman/k8sfake/main.go:113 +0xd20
exit status 2
```

The expected result was that if it worked with a real cluster, it should work with the fake client.

### PR DESCRIPTION
**Pull request by liorokman**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

Make a best effort to translate objects stored in the tracker as concrete
object types and not Unstructured types.

If the object GVR is registered in the provided schema, then it is translated
to a concrete object type (not Unstructured) when it is added into the tracker.

This allows future strategic patch attempts to work without failing on
the 'unable to find api field in struct Unstructured for the json field' error. This
error is caused by the strategic patch code trying to find struct tags on the
Unstructured object in the tracker which match the provided JSON document, but failing.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/client-go/issues/613

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 1]
```diff
@@ -367,6 +368,19 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 	// copy.
 	obj = obj.DeepCopyObject()
 
+	if unstruct, ok := obj.(*unstructured.Unstructured); ok {
+		// Make a best-effort to translate the provided Unstructured object to a concrete type. If successful, then
+		// strategic patches will work on this instance, otherwise they will fail.
+		kinds, _, err := t.scheme.ObjectKinds(obj)
+		if err == nil && len(kinds) > 0 {
+			if newObj, err := t.scheme.New(kinds[0]); err == nil {
+				if err = runtime.DefaultUnstructuredConverter.FromUnstructured(unstruct.Object, newObj); err == nil {
```
Comment (yue9944882) [inline staging/src/k8s.io/client-go/testing/fixture.go]: it should be only working for patch, or odds things will happen like, if you create w/ an unstructured but get a typed instance returned, which is, breaking more essential behavior of the fake client. additionally, i think the conversion can't mock out apiserver's behavior, encoding by unstructured-scheme and then decoding using the provided scheme will be better.

Comment (liorokman) [reply staging/src/k8s.io/client-go/testing/fixture.go]: I checked the fake client code with the following helper program:

```golang
	var unst unstructured.Unstructured
	unst.UnmarshalJSON([]byte(`{"apiVersion": "v1", "kind": "Secret", "type": "Opaque", "metadata": {"name": "test", "namespace": "testns"}, "data": {"test": "dGVzdAo="} }`))

	client.CoreV1().Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "testns"}})
	dynSecret := dyn.Resource(schema.GroupVersionResource{Group: "", Version: "v1", Resource: "secrets"})
	if _, e := dynSecret.Namespace("testns").Create(&unst, metav1.CreateOptions{}); e != nil {
		panic(e)
	}
	if us, e := dynSecret.Namespace("testns").Get("test", metav1.GetOptions{}); e != nil {
		panic(e)
	} else {
  	       fmt.Printf("Unstructured Secret is %#v\n", us)
        }
	if s, e := client.CoreV1().Secrets("testns").Get("test", metav1.GetOptions{}); e != nil {
		panic(e)
	} else {
	     fmt.Printf("Secret is %#v\n", s)
        }
```
The original fake client (before and after my suggested patch) crashes when trying to retrieve the secret as a `v1.Secret` object. I think that this is a different bug altogether.

Looking at the API provided by the client, if you use the typed clientset to retrieve an instance, then the fake client must return a typed instance, regardless of how that instance was created. This is exactly how the API is defined, so doing anything else is a bug - the fake client wouldn't behave like the thing it's trying to fake. I don't think it's essential behavior for the fake client to support a flow that doesn't work with the real apiserver.

The reverse is also impossible: regardless of how you created the object (with a typed instance or an Unstructured one), if you use the dynamic client you expect an `Unstructured` object to be returned. The code even does this translation in the dynamic simple client.

I looked at doing the translation only before the strategic patch, but the code doesn't have access to the ObjectScheme there, so I couldn't actually perform the conversion.

I'm new to the Kubernetes object model - I don't follow what you mean by encoding by the Unstructured-scheme and decoding with the provided scheme. Could you show me an example of this, or a link to the correct API?

**Comment by deads2k** [comment]
Part of the point of the dynamic client was to avoid conversions like this.  Strategic merge patch in particular has been problematic over time and isn't supported for the CRDs.  Since more resources do not work with strategic merge patch than do and we're moving in a direction of more focus on server-side apply, I don't think we need to move in a direction to support this feature.  It's very confusing to users when a client sends a patch that the server rejects on a type basis.

**Comment by liorokman** [comment]
> Strategic merge patch in particular has been problematic over time and isn't supported for the CRDs.

In my use-case, I have a controller which creates standard (corev1, not CRD) resources using the dynamic interface, and then later on patches them as required. When working with the real API server, this works fine. My problem is that the Fake dynamic clientset simply crashes on the patch operation, and this is due to a type mismatch which is even harder to understand - the infamous "unable to find api field in struct Unstructured for the json field 'metadata'". The underlying problem is that the strategic merge code relies on the golang structure tags which don't exist for the Unstructured struct.

> ...  we're moving in a direction of more focus on server-side apply ...

This is fine - but then the question becomes how should these things be tested? Does this mean that in order to test scenarios where strategic patch is used on objects created with the dynamic interface, a fake client can't be used?

**Comment by liggitt** [comment]
> Does this mean that in order to test scenarios where strategic patch is used on objects created with the dynamic interface, a fake client can't be used?

Generally, yes. Patch application (especially of more complex patch types like strategic-merge-patch and server-side-apply) is implemented inside the API server. Trying to reimplement it inside a fake client is a non-goal.

**Comment by liorokman** [comment]
Thanks!

Just to verify though - the strategic patch functionality in the real API server is here to stay?

**Comment by liggitt** [comment]
> Just to verify though - the strategic patch functionality in the real API server is here to stay?

yes, that API will be supported per our deprecation policy https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api

**Pull request closed without merging**